File: packages/sdk-python/rigour_qa/scene/schema.py
"""Core Scene data model for Rigour QA."""

from datetime import datetime
from enum import Enum
from typing import Any, Optional
from uuid import uuid4

from pydantic import BaseModel, Field


class Priority(str, Enum):
    """Priority levels for scenes."""

    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class AssertionType(str, Enum):
    """Types of assertions supported by Rigour QA."""

    STATUS_CODE = "status_code"
    BODY_CONTAINS = "body_contains"
    BODY_SCHEMA = "body_schema"
    DB_STATE = "db_state"
    RESPONSE_TIME = "response_time"
    HEADER_CONTAINS = "header_contains"
    SEMANTIC = "semantic"
    CUSTOM = "custom"


class AuthConfig(BaseModel):
    """Authentication configuration."""

    type: str = Field(
        description="Auth type: bearer, basic, api_key, oauth"
    )
    value: Optional[str] = Field(None, description="Auth token/key value")
    username: Optional[str] = None
    password: Optional[str] = None
    client_id: Optional[str] = None
    client_secret: Optional[str] = None

    model_config = {
        "extra": "allow"
    }


class Actor(BaseModel):
    """Represents an actor (user/client) performing actions in a scene."""

    role: str = Field(description="Actor role: admin, user, anonymous, api_client")
    auth: Optional[AuthConfig] = Field(None, description="Authentication config")
    persona: Optional[str] = Field(
        None, description="Natural language description of this actor"
    )
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = {
        "use_enum_values": False
    }


class Step(BaseModel):
    """Represents a single step in a test scenario."""

    action: str = Field(
        description="Action to perform: HTTP request, UI interaction, or delay"
    )
    input: Optional[dict[str, Any]] = Field(
        None, description="Input data for this step (query params, body, etc.)"
    )
    expect: Optional[str] = Field(
        None, description="Expected outcome or assertion for this step"
    )
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = {
        "use_enum_values": False
    }


class Assertion(BaseModel):
    """Represents an expected assertion/outcome."""

    type: AssertionType = Field(description="Type of assertion")
    target: str = Field(description="Target of assertion: response, DB, header, etc.")
    expected: Any = Field(default=None, description="Expected value or condition")
    semantic_prompt: Optional[str] = Field(
        None, description="Custom agent prompt for semantic assertions"
    )
    tolerance: Optional[float] = Field(
        None, description="Tolerance for numeric assertions (e.g., response time)"
    )
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = {
        "use_enum_values": False
    }


class Scene(BaseModel):
    """Core Scene model - a test scenario described in natural or structured language."""

    id: str = Field(default_factory=lambda: str(uuid4()), description="Unique scene ID")
    title: str = Field(description="Short title for the scene")
    description: str = Field(
        description="Natural language or structured description of the test scenario"
    )
    actor: Optional[Actor] = Field(None, description="The actor performing actions")
    steps: Optional[list[Step]] = Field(
        default_factory=list, description="Explicit test steps (optional, can be auto-generated)"
    )
    assertions: Optional[list[Assertion]] = Field(
        default_factory=list, description="Expected outcomes"
    )
    edge_cases: Optional[list[str]] = Field(
        default_factory=list, description="Hints for edge case exploration"
    )
    tags: list[str] = Field(default_factory=list, description="Tags for filtering")
    priority: Priority = Field(
        default=Priority.MEDIUM, description="Priority level"
    )
    created_at: datetime = Field(default_factory=datetime.utcnow)
    metadata: dict[str, Any] = Field(default_factory=dict, description="Arbitrary extra context")

    model_config = {
        "use_enum_values": False,
        "json_schema_extra": {
            "example": {
                "title": "Login with valid credentials",
                "description": "User logs in with valid credentials and sees dashboard",
                "actor": {
                    "role": "user",
                    "persona": "Regular registered user"
                },
                "assertions": [
                    {
                        "type": "status_code",
                        "target": "response",
                        "expected": 200
                    }
                ],
                "priority": "high",
                "tags": ["auth", "critical-path"]
            }
        }
    }

    def add_step(self, action: str, input_data: Optional[dict] = None, expect: Optional[str] = None) -> "Scene":
        """Fluent API to add a step."""
        step = Step(action=action, input=input_data, expect=expect)
        self.steps.append(step)
        return self

    def add_assertion(
        self,
        assertion_type: AssertionType,
        target: str,
        expected: Any,
        semantic_prompt: Optional[str] = None,
    ) -> "Scene":
        """Fluent API to add an assertion."""
        assertion = Assertion(
            type=assertion_type,
            target=target,
            expected=expected,
            semantic_prompt=semantic_prompt,
        )
        self.assertions.append(assertion)
        return self

    def add_edge_case(self, description: str) -> "Scene":
        """Fluent API to add an edge case hint."""
        self.edge_cases.append(description)
        return self

File: packages/sdk-node/src/scene/schema.ts
/**
 * Core Scene data model for Rigour QA.
 * Matches the Python SDK structure with Zod validation.
 */

import { z } from "zod";
import { v4 as uuidv4 } from "uuid";

/**
 * Priority levels for scenes
 */
export enum Priority {
  CRITICAL = "critical",
  HIGH = "high",
  MEDIUM = "medium",
  LOW = "low",
}

/**
 * Types of assertions supported by Rigour QA
 */
export enum AssertionType {
  STATUS_CODE = "status_code",
  BODY_CONTAINS = "body_contains",
  BODY_SCHEMA = "body_schema",
  DB_STATE = "db_state",
  RESPONSE_TIME = "response_time",
  HEADER_CONTAINS = "header_contains",
  SEMANTIC = "semantic",
  CUSTOM = "custom",
}

/**
 * Authentication configuration schema
 */
const AuthConfigSchema = z.object({
  type: z.enum(["bearer", "basic", "api_key", "oauth"]),
  value: z.string().optional(),
  username: z.string().optional(),
  password: z.string().optional(),
  client_id: z.string().optional(),
  client_secret: z.string().optional(),
  extra: z.record(z.unknown()).optional(),
});

export type AuthConfig = z.infer<typeof AuthConfigSchema>;

/**
 * Actor schema - represents a user/client performing actions
 */
const ActorSchema = z.object({
  role: z.enum(["admin", "user", "anonymous", "api_client"]),
  auth: AuthConfigSchema.optional(),
  persona: z.string().optional(),
  metadata: z.record(z.unknown()).default({}),
});

export type Actor = z.infer<typeof ActorSchema>;

/**
 * Step schema - represents a single action in a test scenario
 */
const StepSchema = z.object({
  action: z.string(),
  input: z.record(z.unknown()).optional(),
  expect: z.string().optional(),
  metadata: z.record(z.unknown()).default({}),
});

export type Step = z.infer<typeof StepSchema>;

/**
 * Assertion schema - represents an expected outcome
 */
const AssertionSchema = z.object({
  type: z.nativeEnum(AssertionType),
  target: z.string(),
  expected: z.unknown(),
  semantic_prompt: z.string().optional(),
  tolerance: z.number().optional(),
  metadata: z.record(z.unknown()).default({}),
});

export type Assertion = z.infer<typeof AssertionSchema>;

/**
 * Scene schema - core model for a test scenario
 */
const SceneSchema = z.object({
  id: z.string().default(() => uuidv4()),
  title: z.string(),
  description: z.string(),
  actor: ActorSchema.optional(),
  steps: z.array(StepSchema).default([]),
  assertions: z.array(AssertionSchema).default([]),
  edge_cases: z.array(z.string()).default([]),
  tags: z.array(z.string()).default([]),
  priority: z.nativeEnum(Priority).default(Priority.MEDIUM),
  created_at: z.date().default(() => new Date()),
  metadata: z.record(z.unknown()).default({}),
});

export type Scene = z.infer<typeof SceneSchema>;

/**
 * Scene class with fluent builder API matching Python SDK
 */
export class SceneClass implements Scene {
  id: string;
  title: string;
  description: string;
  actor?: Actor;
  steps: Step[];
  assertions: Assertion[];
  edge_cases: string[];
  tags: string[];
  priority: Priority;
  created_at: Date;
  metadata: Record<string, unknown>;

  constructor(data: Partial<Scene> & { title: string; description: string }) {
    const validated = SceneSchema.parse(data);
    this.id = validated.id;
    this.title = validated.title;
    this.description = validated.description;
    this.actor = validated.actor;
    this.steps = validated.steps;
    this.assertions = validated.assertions;
    this.edge_cases = validated.edge_cases;
    this.tags = validated.tags;
    this.priority = validated.priority;
    this.created_at = validated.created_at;
    this.metadata = validated.metadata;
  }

  /**
   * Fluent API to add a step
   */
  addStep(
    action: string,
    input?: Record<string, unknown>,
    expect?: string
  ): this {
    const step = StepSchema.parse({ action, input, expect });
    this.steps.push(step);
    return this;
  }

  /**
   * Fluent API to add an assertion
   */
  addAssertion(
    type: AssertionType,
    target: string,
    expected: unknown,
    semanticPrompt?: string
  ): this {
    const assertion = AssertionSchema.parse({
      type,
      target,
      expected,
      semantic_prompt: semanticPrompt,
    });
    this.assertions.push(assertion);
    return this;
  }

  /**
   * Fluent API to add an edge case hint
   */
  addEdgeCase(description: string): this {
    this.edge_cases.push(description);
    return this;
  }

  /**
   * Fluent API to add a tag
   */
  addTag(tag: string): this {
    this.tags.push(tag);
    return this;
  }

  /**
   * Fluent API to set priority
   */
  setPriority(priority: Priority): this {
    this.priority = priority;
    return this;
  }

  /**
   * Fluent API to set actor
   */
  setActor(actor: Actor): this {
    this.actor = actor;
    return this;
  }

  /**
   * Validate the scene
   */
  validate(): boolean {
    SceneSchema.parse(this);
    return true;
  }

  /**
   * Convert to plain object
   */
  toJSON(): Scene {
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      actor: this.actor,
      steps: this.steps,
      assertions: this.assertions,
      edge_cases: this.edge_cases,
      tags: this.tags,
      priority: this.priority,
      created_at: this.created_at,
      metadata: this.metadata,
    };
  }
}

/**
 * Export Zod schemas for runtime validation
 */
export const Schemas = {
  AuthConfig: AuthConfigSchema,
  Actor: ActorSchema,
  Step: StepSchema,
  Assertion: AssertionSchema,
  Scene: SceneSchema,
};
